javascript/ql/lib/semmle/javascript/frameworks/Spife.qll

```diff
@@ -0,0 +1,412 @@
+     */
+    DataFlow::ParameterNode getRequestParameter() { result = this.getParameter(0) }
+
+    /**
+     * Gets the parameter of the route handler that contains the context object.
+     */
+    DataFlow::ParameterNode getContextParameter() { result = this.getParameter(1) }
+  }
+
+  /**
+   * A standard Spife route handler.
+   */
+  private class StandardRouteHandler extends RouteHandler, DataFlow::FunctionNode {
+    StandardRouteHandler() { any(RouteSetup setup).getARouteHandler() = this }
+  }
+
+  /**
+   * A function that looks like a Spife route handler.
+   *
+   * For example, this could be the function `function(req, res, next){...}`.
+   */
+  class RouteHandlerCandidate extends Http::RouteHandlerCandidate {
+    RouteHandlerCandidate() {
+      // heuristic: parameter names match the Restify documentation
+      astNode.getNumParameter() = 2 and
+      astNode.getParameter(0).getName() = ["request", "req"] and
+      astNode.getParameter(1).getName() = ["context", "ctx"]
+    }
+  }
+
+  /**
+   * A Spife request source, that is, the request parameter of a
+   * route handler.
+   */
+  private class RequestSource extends Http::Servers::RequestSource {
+    RouteHandler rh;
+
+    RequestSource() { this = rh.getRequestParameter() }
+
+    /**
+     * Gets the route handler that handles this request.
+     */
+    override RouteHandler getRouteHandler() { result = rh }
+  }
+
+  /**
+   * A Spife context source, that is, the context the parameter of a
+   * route handler.
+   */
+  private class ContextSource extends Http::Servers::RequestSource {
+    RouteHandler rh;
+
+    ContextSource() { this = rh.getContextParameter() }
+
+    /**
+     * Gets the route handler that handles this request.
+     */
+    override RouteHandler getRouteHandler() { result = rh }
+  }
+
+  /**
+   * An access to a user-controlled Spife request input.
+   */
+  private class RequestInputAccess extends Http::RequestInputAccess {
+    RouteHandler rh;
+    string kind;
+
+    RequestInputAccess() {
+      this = rh.getARequestSource().ref().getAPropertyRead("body") and
+      kind = "body"
+      or
+      this = rh.getARequestSource().ref().getAPropertyRead("query").getAPropertyRead() and
+      kind = "parameter"
+      or
+      this = rh.getARequestSource().ref().getAPropertyRead("raw") and
+      kind = "raw"
+      or
+      this = rh.getARequestSource().ref().getAPropertyRead(["url", "urlObject"]) and
+      kind = "url"
+      or
+      this = rh.getARequestSource().ref().getAMethodCall() and
+      this.(DataFlow::MethodCallNode).getMethodName() = ["cookie", "cookies"] and
+      kind = "cookie"
+      or
+      exists(DataFlow::PropRead validated, DataFlow::MethodCallNode get |
+        rh.getARequestSource().ref().getAPropertyRead() = validated and
+        validated.getPropertyName().matches("validated%") and
+        get.getReceiver() = validated and
+        this = get and
+        kind = "body"
+      )
```

Comment:
It's hard to read what this does, and I suspect there might be an issue here.  
But without some example code, that's hard to tell.   

Could you add a comment with an example of what this will match?

Reply:
sure, will do